File: Programmerare.CrsTransformations.Test/crsTransformations/compositeTransformations/CrsTransformationAdapterWeightTest.java
package com.programmerare.crsTransformations.compositeTransformations;

import static org.junit.jupiter.api.Assertions.*;

import com.programmerare.crsTransformationAdapterGeoTools.CrsTransformationAdapterGeoTools;
import com.programmerare.crsTransformations.CrsTransformationAdapter;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;


class CrsTransformationAdapterWeightTest {

    private CrsTransformationAdapter crsTransformationAdapterInstanceNotNull;
    
    @BeforeEach
    void setup() {
        crsTransformationAdapterInstanceNotNull = new CrsTransformationAdapterGeoTools();    
    }
    
    @Test
    void createFromInstance_shouldThrowException_whenAdapterParameterIsNull() {
        assertThrows(
            IllegalArgumentException.class,
            () -> CrsTransformationAdapterWeight.createFromInstance(
                null, // adapter
                123 // weight
            )
        );
    }

    @Test
    void createFromInstance_shouldThrowException_whenWeightParameterIsNegative() {
        assertThrows(
            IllegalArgumentException.class,
            () -> CrsTransformationAdapterWeight.createFromInstance(
                crsTransformationAdapterInstanceNotNull,
                -1 // weight.  null weight leads to compiler error so that it imposslble
            )
        );
    }

    @Test
    void createFromInstance_shouldThrowException_whenWeightParameterIsZero() {
        assertThrows(
            IllegalArgumentException.class,
            () -> CrsTransformationAdapterWeight.createFromInstance(
                crsTransformationAdapterInstanceNotNull,
                0 // weight.  null weight leads to compiler error so that it imposslble
            )
        );
    }

    @Test
    void createFromStringWithFullClassNameForImplementation_shouldThrowException_whenTheClassIsNotImplementingTheExpectedInterface() {
        final String nameOfClassNotImplementingTheInterfaceCrsTransformationAdapter = this.getClass().getName();
        assertThrows(
            Exception.class, // not testing exactly what exception type but just some class inheriting from Exception 
            () -> CrsTransformationAdapterWeight.createFromStringWithFullClassNameForImplementation(
                nameOfClassNotImplementingTheInterfaceCrsTransformationAdapter,
                123 // weight
            )
        );        
    }

    @Test
    void createFromStringWithFullClassNameForImplementation_shouldSucceed_whenTheClassIsImplementingTheExpectedInterface() {
        final String nameOfClassImplementingTheInterfaceCrsTransformationAdapter = crsTransformationAdapterInstanceNotNull.getClass().getName();
        final double weightValue = 123;
        CrsTransformationAdapterWeight crsTransformationAdapterWeight = CrsTransformationAdapterWeight.createFromStringWithFullClassNameForImplementation(
            nameOfClassImplementingTheInterfaceCrsTransformationAdapter,
            weightValue
        );
        assertNotNull(crsTransformationAdapterWeight);
        assertEquals(
            nameOfClassImplementingTheInterfaceCrsTransformationAdapter,  
            crsTransformationAdapterWeight.getCrsTransformationAdapter().getClass().getName()
        );
        assertEquals(
            weightValue, // expected
            crsTransformationAdapterWeight.getWeight()
        );
    }

}

File: Programmerare.CrsTransformations.Test/CrsTransformations/CompositeTransformations/CrsTransformationAdapterWeightTest.cs
namespace Programmerare.CrsTransformations.CompositeTransformations {

using System;
using NUnit.Framework;
using Programmerare.CrsTransformations;
using Programmerare.CrsTransformations.Adapter.MightyLittleGeodesy;

class CrsTransformationAdapterWeightTest {

    private ICrsTransformationAdapter crsTransformationAdapterInstanceNotNull;
    private CrsTransformationAdapterWeightFactory weightFactory;
        
    [SetUp]
    public void setup() {
        weightFactory = CrsTransformationAdapterWeightFactory.Create();
        crsTransformationAdapterInstanceNotNull = new CrsTransformationAdapterMightyLittleGeodesy();
    }
    
    [Test]
    public void createFromInstance_shouldThrowException_whenAdapterParameterIsNull() {
        ArgumentNullException exception = Assert.Throws<ArgumentNullException>(() => {
            weightFactory.CreateFromInstance(
                null, // adapter
                123 // weight
            );
        });
    }

    [Test]
    public void createFromInstance_shouldThrowException_whenWeightParameterIsNegative() {
        ArgumentException exception = Assert.Throws<ArgumentException>(() => {
            weightFactory.CreateFromInstance(
                crsTransformationAdapterInstanceNotNull,
                -1 // weight.  null weight leads to compiler error so that it imposslble
            );
        });
    }

    [Test]
    public void createFromInstance_shouldThrowException_whenWeightParameterIsZero() {
        ArgumentException exception = Assert.Throws<ArgumentException>(() => {
            weightFactory.CreateFromInstance(
                crsTransformationAdapterInstanceNotNull,
                0.0 // weight.  null weight leads to compiler error so that it imposslble
            );
        });
    }

    [Test]
    public void createFromStringWithFullClassNameForImplementation_shouldThrowException_whenTheClassIsNotImplementingTheExpectedInterface() {
        string nameOfClassNotImplementingTheInterfaceCrsTransformationAdapter = this.GetType().FullName;
        ArgumentException exception = Assert.Throws<ArgumentException>(() => {
            weightFactory.CreateFromStringWithFullClassNameForImplementation(
                nameOfClassNotImplementingTheInterfaceCrsTransformationAdapter,
                123 // weight
            );
        });
    }

    [Test]
    public void createFromStringWithFullClassNameForImplementation_shouldSucceed_whenTheClassIsImplementingTheExpectedInterface() {
        string nameOfClassImplementingTheInterfaceCrsTransformationAdapter = crsTransformationAdapterInstanceNotNull.GetType().FullName;
        double weightValue = 123;
        CrsTransformationAdapterWeight crsTransformationAdapterWeight = weightFactory.CreateFromStringWithFullClassNameForImplementation(
            nameOfClassImplementingTheInterfaceCrsTransformationAdapter,
            weightValue
        );
        Assert.IsNotNull(crsTransformationAdapterWeight);
        Assert.AreEqual(
            nameOfClassImplementingTheInterfaceCrsTransformationAdapter,  
            crsTransformationAdapterWeight.CrsTransformationAdapter.GetType().FullName
        );
        Assert.AreEqual(
            weightValue, // expected
            crsTransformationAdapterWeight.Weight
        );
    }

}
}